### PR TITLE
Feature/error email generation

### PR DESCRIPTION
**Pull request by asktheaxis**
Action for building the execution tree is complete. Moving on to finding/formatting error

### CODE REVIEW

File: actions/get_formatted_error.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        # python actions ex. (vsphere pack)
+        if 'stderr' in error_result:
+            return error_result['stderr']
+
+        return "Could not retrieve error message"
+
+    def run(self, **kwargs):
+
+        st2_exe_id = kwargs['st2_exe_id']
+        cmdb_request_item_url = kwargs['cmdb_request_item_url']
```
Comment (bishopbm1) [inline actions/get_formatted_error.py]: lets not have the cmdb data be apart of this. the cmdb side of things is something encore specific and i think its probably not something alot of other people would use and if they do want it is easy enough to add on their own

File: README.md [hunk 2]
```diff
@@ -1,16 +1,66 @@
-# errors Integration Pack
+[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
 
-## Configuration
-TODO: Describe configuration
+# Errors Pack
 
+# Introduction
+This pack holds all of the contains all actions and workflows needed for finding and formating
```
Comment (bishopbm1) [inline README.md]: wording is a little messed up here. Lets give some information about how we get the error message and tree using the st2 api (python module)

File: README.md [hunk 3]
```diff
@@ -1,16 +1,66 @@
...
+    ```
+
+# Usage
+
+## Actions
+
+| Action | Description |
+|--------|-------------|
+| build_execution_tree | Builds an execution tree of the given task |
+| get_formatted_error | Finds an error in the given task and formats the result into an HTML tagged output |
```
Comment (bishopbm1) [inline README.md:35]: need to add your workflow here

File: README.md [hunk 4]
```diff
@@ -1,16 +1,66 @@
...
+
+### Action Example - errors.build_execution_tree
+
+`errors.build_execution_tree` builds an execution tree of a given StackStorm task:
```
Comment (bishopbm1) [inline README.md:40]: For all of these give a sample output so the user knows what to expect and how to handle it in their workflows

File: actions/get_formatted_error.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from six import string_types
+from lib.base_action import BaseAction
+
+
+IGNORED_ERROR_TASKS = [
```
Comment (bishopbm1) [inline actions/get_formatted_error.py]: These would be encore specific ignored tasks. We need to make this an input and pass it to the function where it is used

File: actions/get_formatted_error.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                self.parent_error = execution  # pylint: disable=no-member
+                if hasattr(execution, 'children'):
+                    for c in execution.children:
+                        self.find_error_execution(c)
+                else:
+                    self.child_error.append(execution)  # pylint: disable=no-member
+
+    def format_error(self, cmdb_request_item_url, cmdb_request_item):
+        err_string = ""
+        if cmdb_request_item_url:
```
Comment (bishopbm1) [inline actions/get_formatted_error.py]: see other comments about now using cmdb stuff in this pack

File: actions/get_formatted_error.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            err_string += ("ServiceNow Request: <a href={0}>{1}</a><br>"
+                            "".format(cmdb_request_item_url, cmdb_request_item))
+
+        if self.child_error:  # pylint: disable=no-member
```
Comment (bishopbm1) [inline actions/get_formatted_error.py]: I think if you initialize these variables in the init function you wouldn't need `# pylint: disable=no-member`

File: actions/get_formatted_error.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            parent_error = self.get_error_message(error_result)
+            err_message = self.format_error_strings(parent_error)
+            err_string += ("Error task: {0}<br>Error execution ID: {1}<br>Error message: {2}"
+                           "<br>".format(self.parent_error.context['orquesta']['task_name'],
+                                         self.parent_error.id,
+                                         err_message))
+
+        return err_string
+
+    def format_error_escaped_returns(self):
```
Comment (bishopbm1) [inline actions/get_formatted_error.py]: This function looks the same as above if the only thing that changes between them is adding new lines vs line breaks then make just that part into a different function.

File: actions/get_formatted_error.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+                    result_set = error_result['result']['details']['result_set'][0]
+                    return result_set['value']['_error']['msg']
+
+            if isinstance(error_result['result'], string_types):
+                return error_result['result']
+
+            if 'stderr' in error_result['result']:
+                return error_result['result']['stderr']
+
+        # python actions ex. (vsphere pack)
```
Comment (bishopbm1) [inline actions/get_formatted_error.py]: lets give a link to the vsphere pack like you did above for bolt

File: actions/workflows/get_error_data.yaml [hunk 10]
```diff
@@ -0,0 +1,84 @@
+---
+version: 1.0
+
+description: Gets the execution tree and the resulting error of a given execution
+
+input:
+  - cmdb_request_item
+  - cmdb_request_item_url
+  - retire
```
Comment (bishopbm1) [inline actions/workflows/get_error_data.yaml]: retirement is something that is encore specific. If we need to do certain things in certain situations we should do it on the encore side. This workflow should be user generic and just run both tasks and output the data for the user to use however they need.

File: actions/workflows/get_error_data.yaml [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  - st2_exe_id
+  - sn_update
```
Comment (bishopbm1) [inline actions/workflows/get_error_data.yaml]: See comment above. I think this is probably used to know if the values should be newlines vs line breaks so we should call it something else so and that can be passed in and the user knows what to expect

File: actions/get_formatted_error.yaml [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+parameters:
+  cmdb_request_item:
+    type: string
+    description: "CMDB request item if the VM was requested through a CMDB"
+    required: false
+  cmdb_request_item_url:
+    type: string
+    description: "CMDB request item url if the VM was requested through a CMDB"
+    required: false
+  sn_update:
```
Comment (bishopbm1) [inline actions/get_formatted_error.yaml]: Lets call this something else to know if we need to use newlines vs line breaks. Calling it `sn_update` doesn't really tell the user what it is doing.

File: actions/get_error_data.yaml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+parameters:
+  cmdb_request_item:
+    type: string
+    description: "CMDB request item if the VM was requested through a CMDB"
+    required: false
+  cmdb_request_item_url:
+    type: string
+    description: "CMDB request item url if the VM was requested through a CMDB"
+    required: false
+  retire:
```
Comment (bishopbm1) [inline actions/get_error_data.yaml]: See comment below

File: actions/get_error_data.yaml [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    type: boolean
+    description: "True if the error is during retirement"
+    required: true
+    default: false
+  sn_update:
```
Comment (bishopbm1) [inline actions/get_error_data.yaml]: See comment below

File: README.md [hunk 15]
```diff
@@ -1,16 +1,66 @@
...
+```yaml
+get_error_execution_tree:
+  action: errors.build_execution_tree
+  input:
+    st2_exe_id: "5fa45525935a74a08162cd7b"
+```
+
+### Action Example - errors.get_formatted_error
+
+`errors.get_formatted_error` Finds an error in the given task and formats the result into an HTML tagged output
```
Comment (bishopbm1) [inline README.md]: Make sure you talk about the differences in the optional formatting

**Review by bishopbm1 — COMMENTED** [review]
Looks good! Just a couple of small things to cleanup

File: actions/build_execution_tree.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+            self.task_list.append(task_dict)  # pylint: disable=no-member
+            if has_child:
+                for c in execution.children:
+                    self.get_execution_tree(c, delimeter + '   ')
+
+        return self.task_list
+
+    def run(self, **kwargs):
+
+        st2_exe_id = kwargs['st2_exe_id']
```
Comment (nmaludy) [inline actions/build_execution_tree.py]: don't need kwargs here, just specify the parameter in the run definition

**Review by nmaludy — CHANGES_REQUESTED** [review]

**Pull request merged**